src/main/java/uk/gov/hmcts/reform/ccd/documentam/service/impl/DocumentManagementServiceImpl.java

```diff
@@ -186,56 +186,72 @@ private boolean allowMetadataOverride(Document document, CaseDocumentsMetadata c
-            caseTypeId = authorisedService.getDefaultCaseTypeForTokenGeneration();
+        Either<ResourceNotFoundException, Document> either = documentStoreClient.getDocument(documentId);
+        if (either.isLeft()) {
+            return "";
+        }
+        Document documentMetaData = either.get();
+
+        boolean metadataPatchRequired = false;
+
+        String caseTypeId;
+        if (StringUtils.isNotEmpty(documentMetaData.getCaseTypeId())) {
+            caseTypeId = documentMetaData.getCaseTypeId();
+        } else {
+            if (authorisedService.getCaseTypeIdOptionalFor().contains(permission)) {
+                caseTypeId = authorisedService.getDefaultCaseTypeForTokenGeneration();
+                metadataPatchRequired = true;
+            } else {
+                throw new ForbiddenException("No case type id available to generate hash token for document "
+                                                 + documentId);
+            }
         }
-        final String finalCaseTypeId = caseTypeId;
 
-        String jurisdictionId = null;
-        if (authorisedService.getJurisdictionIdOptionalFor().contains(permission)) {
-            jurisdictionId = authorisedService.getDefaultJurisdictionForTokenGeneration();
+        String jurisdictionId;
+        if (StringUtils.isNotEmpty(documentMetaData.getJurisdictionId())) {
+            jurisdictionId = documentMetaData.getJurisdictionId();
+        } else {
+            if (authorisedService.getJurisdictionIdOptionalFor().contains(permission)) {
+                jurisdictionId = authorisedService.getDefaultJurisdictionForTokenGeneration();
+                metadataPatchRequired = true;
+            } else {
+                throw new ForbiddenException("No jurisdiction id available to generate hash token for document "
+                                                 + documentId);
+            }
         }
 
-        final String finalJurisdictionId = jurisdictionId;
+        String hashToken = generateHashToken(documentId, documentMetaData.getCaseId(), jurisdictionId, caseTypeId);
 
-        return documentStoreClient.getDocument(documentId)
-            .fold(
-                resourceNotFound -> "",
-                document -> generateHashToken(documentId, document, finalCaseTypeId, finalJurisdictionId)
+        if (metadataPatchRequired) {
+            List<DocumentUpdate> documentsList = new ArrayList<>();
+            final DocumentUpdate documentUpdate = new DocumentUpdate(
+                documentId,
+                Map.of(METADATA_CASE_TYPE_ID, caseTypeId,
+                    METADATA_JURISDICTION_ID, jurisdictionId)
             );
+
+            documentsList.add(documentUpdate);
+            UpdateDocumentsCommand updateDocumentsCommand = new UpdateDocumentsCommand(NULL_TTL, documentsList);
+            documentStoreClient.patchDocumentMetadata(updateDocumentsCommand);
+        }
```

Comment:
this would better be inside a separate method.

Reply:
done